code/lib/core-server/src/utils/StoryIndexGenerator.ts

```diff
@@ -250,19 +251,98 @@ export class StoryIndexGenerator {
     );
   }
 
-  async extractStories(specifier: NormalizedStoriesSpecifier, absolutePath: Path) {
+  async extractStories(
+    specifier: NormalizedStoriesSpecifier,
+    absolutePath: Path
+  ): Promise<StoriesCacheEntry | DocsCacheEntry> {
     const relativePath = path.relative(this.options.workingDir, absolutePath);
-    const entries = [] as IndexEntry[];
     const importPath = slash(normalizeStoryPath(relativePath));
-    const makeTitle = (userTitle?: string) => {
+    const defaultMakeTitle = (userTitle?: string) => {
       return userOrAutoTitleFromSpecifier(importPath, specifier, userTitle);
     };
 
-    const storyIndexer = this.options.storyIndexers.find((ind) => ind.test.exec(absolutePath));
-    if (!storyIndexer) {
+    const indexer = (this.options.indexers as StoryIndexer[])
+      .concat(this.options.storyIndexers)
+      .find((ind) => ind.test.exec(absolutePath));
+
+    if (!indexer) {
       throw new Error(`No matching indexer found for ${absolutePath}`);
     }
-    const csf = await storyIndexer.indexer(absolutePath, { makeTitle });
+    if (indexer.indexer) {
+      return this.extractStoriesFromDeprecatedIndexer({
+        indexer: indexer.indexer,
+        indexerOptions: { makeTitle: defaultMakeTitle },
+        absolutePath,
+        importPath,
+      });
+    }
+
+    const indexInputs = await indexer.index(absolutePath, { makeTitle: defaultMakeTitle });
+
+    const entries: (StoryIndexEntry | DocsCacheEntry)[] = indexInputs.map((input) => {
+      const name = input.name ?? storyNameFromExport(input.key);
+      const title = input.title ?? defaultMakeTitle();
+      const id = input.id ?? toId(title, name);
+      const tags = (input.tags || []).concat('story');
+
+      return {
+        type: 'story',
+        id,
+        name,
+        title,
+        importPath,
+        tags,
+      };
+    });
+
+    const { autodocs } = this.options.docs;
+    // We need a docs entry attached to the CSF file if either:
+    //  a) autodocs is globally enabled
+    //  b) we have autodocs enabled for this file
+    //  c) it is a stories.mdx transpiled to CSF
+    const hasAutodocsTag = entries.some((entry) => entry.tags.includes(AUTODOCS_TAG));
+    const isStoriesMdx = entries.some((entry) => entry.tags.includes(STORIES_MDX_TAG));
+    const createDocEntry =
+      autodocs === true || (autodocs === 'tag' && hasAutodocsTag) || isStoriesMdx;
+
+    if (createDocEntry) {
+      const name = this.options.docs.defaultName;
+      // TODO: how to get "component title" or "component tags" when we only have direct stories here?
```

Comment:
as we don't have a concept of `component` but only story entries here now, I don't see a way to get tags and title from the meta. I just opted to get it from the first entry.
If anyone can come up with a smarter way or a reason why this is bad, let me know.